### PR TITLE
Add indexOfAssumeSorted Function. request #17795

### PR DESCRIPTION
**Pull request by erickurbanov**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add indexOfAssumeSorted function for array types. Optimizes the search in the case of a sorted in non-decreasing order array.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

```sql
CREATE test_table(id Uint64, numbers Array(Int64));
SELECT indexOfAssumeSorted(numbers, x) FROM test_table;
```
indexOfAssumeSorted(Array, x)
Use only if the array is sorted, the function returns the index of the first occurrence of the element x

Implementation of #17795

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

File: src/Functions/array/arrayIndex.h [hunk 1]
```diff
@@ -53,6 +53,10 @@ struct IndexOfAction
     static constexpr void apply(ResultType& current, size_t j) noexcept { current = j + 1; }
 };
 
+struct IndexOfAssumeSorted : public IndexOfAction
```
Comment (erickurbanov) [inline src/Functions/array/arrayIndex.h:56]: for constexpr purpose

**Comment by alexey-milovidov** [comment]
Let's check the performance benefits.

**Comment by erickurbanov** [comment]
@alexey-milovidov https://github.com/ClickHouse/ClickHouse/commit/eb1cce7b07bae5dcb80868f449db1b4dd7dbeb52#diff-02fe639702b825f133af028b929196d687488ca5068f1fd215f2ef08c2048158L1
this test works with this option set allow_experimental_analyzer=0;
ClickHouse/tests/queries/0_stateless/02944_variant_as_common_type.sql

File: docs/en/sql-reference/functions/array-functions.md [hunk 2]
```diff
@@ -786,6 +786,24 @@ SELECT indexOf([1, 3, NULL, NULL], NULL)
...
+If the internal array type is Nullable, the ‘indexOf‘ function will be used.
+
+Example:
+
+``` sql
+SELECT indexOfAssumeSorted([1, 3, 3, 3, 4, 4, 5], 4)
+```
+
+``` text
+┌─indexOf([1, 3, 3, 3, 4, 4, 5], NULL)─┐
```
Comment (erickurbanov) [inline docs/en/sql-reference/functions/array-functions.md:802]: wrong example, i`ll fix it here and further

**Comment by erickurbanov** [comment]
@alexey-milovidov The checks are green now. I noticed that there are changes in perfomance. How accurate are these tests? Do I need to look at perf data ?

**Comment by erickurbanov** [comment]
@alexey-milovidov gentle ping

**Comment by alexey-milovidov** [comment]
Thanks! How does the performance compare?

**Comment by erickurbanov** [comment]
> Thanks! How does the performance compare?

@alexey-milovidov for example https://s3.amazonaws.com/clickhouse-test-reports/72517/83adb29584fdf7ae482b3ad5cfa00c8996da93d2/performance_comparison__release__[4_4]/report.html

I do not understand well whether my changes could have affected the loss of performance in some cases.
I did not find detailed documentation in clickhouse about perfomance tests.
Can you please explain to me if everything is good in the tests?) If everything is fine, I would ask you to review pull request. Thanks.

**Comment by alexey-milovidov** [comment]
The code is reasonable (but I didn't review yet) and the tests are passing.
The difference in performance tests is inside the allowed threshold, and you didn't change any code related to the existing performance tests. The existing performance tests don't test the newly added function.

What I'm interested in is how this new function shows its performance in practice.

**Comment by erickurbanov** [comment]
> The code is reasonable (but I didn't review yet) and the tests are passing. The difference in performance tests is inside the allowed threshold, and you didn't change any code related to the existing performance tests. The existing performance tests don't test the newly added function, so they cannot show anything related to it.
> 
> What I'm interested in is how this new function shows its performance in practice.

@alexey-milovidov I wrote a simple python script for the test. I'm attaching it to the message. https://pastebin.com/DabY5eWM
My setup: Laptop M2 PRO (ARM).
RAM: 32GB
Memory Bandwith: 200GB/s
Cores: 12
Cheap: Apple M2 Pro
L1 cache: 65536 bytes
L2 cache: 37748736 bytes

We have a table size of 50k rows.
Array sizes = {100, 1000, 10000, 100000}.
10 search iterations for each size.
Before each launch of the index search function, we drop caches.
Comparison:
| size | indexOf elapsed | indexOfAssumeSorted  elapsed |
|----------|----------|----------|
| 100   | 0.9059079999999999 sec | 0.8722555999999999  sec |
| 1000    | 0.8636334 sec  | 0.9614133  sec |
| 10000    | 0.9485091000000001 sec | 0.8982650999999999  sec |
| 100000    | 1.2407332  sec | 0.8923985999999999  sec |

obviously, binary search can lose out to linear search on small sizes, but then we have an increase in performance.

**Comment by erickurbanov** [comment]
@alexey-milovidov if necessary to make a benchmark with other parameters, please tell me :)

**Comment by yohannj** [comment]
> @alexey-milovidov if necessary to make a benchmark with other parameters, please tell me :)

Hi @erickurbanov , was just looking at this issue.
Imho, it would be interesting to do the bench directly in ClickHouse, using [clickhouse-benchmark](https://clickhouse.com/docs/en/operations/utilities/clickhouse-benchmark) to not take into account network latency.

You could also disable some clickhouse caches to show the perf one should expect on cold storage.
use_query_cache=false
min_bytes_to_use_direct_io=1 

I think there was another setting, but I can't find it. :)

**Comment by erickurbanov** [comment]
@alexey-milovidov ping

**Review by vdimir — APPROVED** [review]
Looks good overall, sorry for delay in review. Let's wait untill full CI run passes (for some reason only subset of checks were run previous commit)

**Comment by erickurbanov** [comment]
> Looks good overall, sorry for delay in review. Let's wait untill full CI run passes (for some reason only subset of checks were run previous commit)

Thanks. All checks are passed. Can you merge ?

File: docs/en/sql-reference/functions/array-functions.md (hunk 2 above)
Comment (rschu1ze) [reply docs/en/sql-reference/functions/array-functions.md:802]: https://github.com/ClickHouse/ClickHouse/pull/73036

File: tests/queries/0_stateless/03276_index_of_assume_sorted.sql [hunk 3]
```diff
@@ -0,0 +1,26 @@
+DROP TABLE IF EXISTS test;
+
+CREATE TABLE test(
+    id UInt64,
+    numbers Array(Int64)
+)
+ENGINE = MergeTree()
+ORDER BY id;
+
+INSERT INTO test VALUES(1, [1, 2, 2, 3, 3, 3, 4, 4, 4, 5, 6, 7]);
```
Comment (rschu1ze) [inline tests/queries/0_stateless/03276_index_of_assume_sorted.sql:10]: The docs make a mention "If the internal array type is Nullable, the ‘indexOf‘ function will be used.". Why is that case not tested here?

**Comment by clickgapai** [comment]
Hi @erickurbanov @vdimir — while reviewing this PR I found the following:

- **Bug (P1):** `indexOfAssumeSorted` returns wrong results for `Array(LowCardinality(String))` due to binary search on dictionary indices — https://github.com/ClickHouse/ClickHouse/issues/104772

Happy to discuss — close anything that's wrong or already addressed.

**Pull request merged**